Image is no longer Send in 0.3.0

Image now uses internal data structures that do not implement Send. This broke upgrading my application from 0.2.5 to 0.3.0 because I was generating an Image in a UI callback then moving it into an `upgrade_in_event_loop` callback. I hacked around this by passing a `SharedPixelBuffer<Rgba8Pixel>` from the UI callback to the event loop then calling `Image::from_rgba8_premultiplied` in the event loop.

```
moire on  upgrade [$!] via 🦀 v1.63.0 took 3s 
 0% ❯ cargo build
   Compiling moire-ui v0.0.0 (/home/be/sw/moire/crates/ui)
error[E0277]: `Rc<RefCell<rctree::NodeData<usvg::NodeKind>>>` cannot be sent between threads safely
   --> crates/ui/src/deck.rs:196:25
    |
196 |             main_window.upgrade_in_event_loop(move |main_window| {
    |                         ^^^^^^^^^^^^^^^^^^^^^ `Rc<RefCell<rctree::NodeData<usvg::NodeKind>>>` cannot be sent between threads safely
    |
    = help: within `i_slint_core::graphics::image::svg::ParsedSVG`, the trait `Send` is not implemented for `Rc<RefCell<rctree::NodeData<usvg::NodeKind>>>`
    = note: required because it appears within the type `rctree::Node<usvg::NodeKind>`
    = note: required because it appears within the type `usvg::Tree`
    = note: required because it appears within the type `i_slint_core::graphics::image::svg::ParsedSVG`
    = note: required because of the requirements on the impl of `Send` for `VRc<OpaqueImageVTable, i_slint_core::graphics::image::svg::ParsedSVG>`
    = note: required because it appears within the type `ImageInner`
    = note: required because it appears within the type `slint::Image`
note: required because it's used within this closure
   --> crates/ui/src/deck.rs:196:47
    |
196 |               main_window.upgrade_in_event_loop(move |main_window| {
    |  _______________________________________________^
197 | |                 // FIXME: Is the deck still supposed to load the track that has now
198 | |                 // become available, i.e. is a load operation for the corresponding
199 | |                 // track uid still pending? Otherwise we should abort this operation
...   |
241 | |                     .expect("GUI to audio channel full!");
242 | |             }).expect("Failed to load file");
    | |_____________^
note: required by a bound in `slint::Weak::<T>::upgrade_in_event_loop`
   --> /home/be/.cargo/registry/src/github.com-1ecc6299db9ec823/i-slint-core-0.3.0/api.rs:673:36
    |
673 |             func: impl FnOnce(T) + Send + 'static,
    |                                    ^^^^ required by this bound in `slint::Weak::<T>::upgrade_in_event_loop`

error[E0277]: `Rc<RefCell<rctree::NodeData<usvg::NodeKind>>>` cannot be shared between threads safely
   --> crates/ui/src/deck.rs:196:25
    |
196 |             main_window.upgrade_in_event_loop(move |main_window| {
    |                         ^^^^^^^^^^^^^^^^^^^^^ `Rc<RefCell<rctree::NodeData<usvg::NodeKind>>>` cannot be shared between threads safely
    |
    = help: within `i_slint_core::graphics::image::svg::ParsedSVG`, the trait `Sync` is not implemented for `Rc<RefCell<rctree::NodeData<usvg::NodeKind>>>`
    = note: required because it appears within the type `rctree::Node<usvg::NodeKind>`
    = note: required because it appears within the type `usvg::Tree`
    = note: required because it appears within the type `i_slint_core::graphics::image::svg::ParsedSVG`
    = note: required because of the requirements on the impl of `Send` for `VRc<OpaqueImageVTable, i_slint_core::graphics::image::svg::ParsedSVG>`
    = note: required because it appears within the type `ImageInner`
    = note: required because it appears within the type `slint::Image`
note: required because it's used within this closure
   --> crates/ui/src/deck.rs:196:47
    |
196 |               main_window.upgrade_in_event_loop(move |main_window| {
    |  _______________________________________________^
197 | |                 // FIXME: Is the deck still supposed to load the track that has now
198 | |                 // become available, i.e. is a load operation for the corresponding
199 | |                 // track uid still pending? Otherwise we should abort this operation
...   |
241 | |                     .expect("GUI to audio channel full!");
242 | |             }).expect("Failed to load file");
    | |_____________^
note: required by a bound in `slint::Weak::<T>::upgrade_in_event_loop`
   --> /home/be/.cargo/registry/src/github.com-1ecc6299db9ec823/i-slint-core-0.3.0/api.rs:673:36
    |
673 |             func: impl FnOnce(T) + Send + 'static,
    |                                    ^^^^ required by this bound in `slint::Weak::<T>::upgrade_in_event_loop`

error[E0277]: `NonNull<i_slint_core::sharedvector::SharedVectorInner<u8>>` cannot be shared between threads safely
   --> crates/ui/src/deck.rs:196:25
    |
196 |             main_window.upgrade_in_event_loop(move |main_window| {
    |                         ^^^^^^^^^^^^^^^^^^^^^ `NonNull<i_slint_core::sharedvector::SharedVectorInner<u8>>` cannot be shared between threads safely
    |
    = help: within `i_slint_core::graphics::image::svg::ParsedSVG`, the trait `Sync` is not implemented for `NonNull<i_slint_core::sharedvector::SharedVectorInner<u8>>`
    = note: required because it appears within the type `SharedVector<u8>`
    = note: required because it appears within the type `SharedString`
    = note: required because it appears within the type `ImageCacheKey`
    = note: required because it appears within the type `i_slint_core::graphics::image::svg::ParsedSVG`
    = note: required because of the requirements on the impl of `Send` for `VRc<OpaqueImageVTable, i_slint_core::graphics::image::svg::ParsedSVG>`
    = note: required because it appears within the type `ImageInner`
    = note: required because it appears within the type `slint::Image`
note: required because it's used within this closure
   --> crates/ui/src/deck.rs:196:47
    |
196 |               main_window.upgrade_in_event_loop(move |main_window| {
    |  _______________________________________________^
197 | |                 // FIXME: Is the deck still supposed to load the track that has now
198 | |                 // become available, i.e. is a load operation for the corresponding
199 | |                 // track uid still pending? Otherwise we should abort this operation
...   |
241 | |                     .expect("GUI to audio channel full!");
242 | |             }).expect("Failed to load file");
    | |_____________^
note: required by a bound in `slint::Weak::<T>::upgrade_in_event_loop`
   --> /home/be/.cargo/registry/src/github.com-1ecc6299db9ec823/i-slint-core-0.3.0/api.rs:673:36
    |
673 |             func: impl FnOnce(T) + Send + 'static,
    |                                    ^^^^ required by this bound in `slint::Weak::<T>::upgrade_in_event_loop`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `moire-ui` due to 3 previous errors
```

Perhaps [cargo-semver-checks](https://predr.ag/blog/toward-fearless-cargo-update/) could have caught this.

--- a/internal/core/graphics/image.rs
+++ b/internal/core/graphics/image.rs
@@ -541,6 +541,26 @@ pub struct LoadImageError(());
 ///
 /// let image = Image::from_rgba8_premultiplied(pixel_buffer);
 /// ```
+///
+/// ### Sending Image to a thread
+///
+/// `Image` is not [`Send`], because it uses internal cache that are local to the Slint thread.
+/// If you want to create image data in a thread and send that to slint, you can construct the
+/// [`SharedPixelBuffer`] in a thread, and send that to slint
+///
+/// ```rust,no_run
+/// # use i_slint_core::graphics::{SharedPixelBuffer, Image, Rgba8Pixel};
+/// std::thread::spawn(move || {
+///     let mut pixel_buffer = SharedPixelBuffer::<Rgba8Pixel>::new(640, 480);
+///     // ... fill the pixel_buffer with data as shown in the previous example ...
+///     slint::invoke_from_event_loop(move || {
+///         // this will run in the Slint's UI thread
+///         let image = Image::from_rgba8_premultiplied(pixel_buffer);
+///         // ... use the image, eg:
+///         // my_ui_handle.upgrade().unwrap().set_image(image);
+///     });
+/// });
+/// ```
 #[repr(transparent)]
 #[derive(Default, Clone, Debug, PartialEq, derive_more::From)]
 pub struct Image(ImageInner);

--- a/internal/core/graphics/image.rs
+++ b/internal/core/graphics/image.rs
@@ -545,8 +545,8 @@ pub struct LoadImageError(());
 /// ### Sending Image to a thread
 ///
 /// `Image` is not [`Send`], because it uses internal cache that are local to the Slint thread.
-/// If you want to create image data in a thread and send that to slint, you can construct the
-/// [`SharedPixelBuffer`] in a thread, and send that to slint
+/// If you want to create image data in a thread and send that to slint, construct the
+/// [`SharedPixelBuffer`] in a thread, and send that to Slint's UI thread.
 ///
 /// ```rust,no_run
 /// # use i_slint_core::graphics::{SharedPixelBuffer, Image, Rgba8Pixel};

